pkg/acquisition/modules/loki/internal/lokiclient/loki_client.go

```diff
@@ -0,0 +1,255 @@
+			resp, err := http.Get(uri)
+
+			if err != nil {
+				return errors.Wrapf(err, "error querying range")
+			}
+			if resp.StatusCode != http.StatusOK {
+				body, _ := io.ReadAll(resp.Body)
+				resp.Body.Close()
+				return errors.Wrapf(err, "bad HTTP response code: %d: %s", resp.StatusCode, string(body))
+			}
+
+			var lq LokiQueryRangeResponse
+
+			json.NewDecoder(resp.Body).Decode(&lq)
+			resp.Body.Close()
+
+			lc.Logger.Tracef("Got response: %+v", lq)
+
+			c <- &lq
+
+			if len(lq.Data.Result) == 0 || len(lq.Data.Result[0].Entries) < lc.config.Limit {
+				lc.Logger.Infof("Got less than %d results (%d), stopping", lc.config.Limit, len(lq.Data.Result))
+				close(c)
+				return nil
+			}
+			//Can we assume we will always have only one stream?
+			lastTs := lq.Data.Result[0].Entries[len(lq.Data.Result[0].Entries)-1].Timestamp
+
+			lc.Logger.Infof("Got %d results, last timestamp: %s (converted: %s)", len(lq.Data.Result[0].Entries), lastTs, strconv.Itoa(lastTs.Nanosecond()))
+			u, err := url.Parse(uri) //we can ignore the error, we know it's valid
+			if err != nil {
+				return errors.Wrapf(err, "error parsing URL")
+			}
+			queryParams := u.Query()
+			queryParams.Set("start", strconv.Itoa(int(lastTs.UnixNano())))
+			u.RawQuery = queryParams.Encode()
+			uri = u.String()
+		}
+	}
+}
+
+func (lc *LokiClient) getURLFor(endpoint string, params map[string]string) string {
+	u, err := url.Parse(lc.config.LokiURL)
+	if err != nil {
+		return ""
+	}
+	queryParams := u.Query()
+	for k, v := range params {
+		queryParams.Set(k, v)
+	}
+	u.RawQuery = queryParams.Encode()
+
+	u.Path, err = url.JoinPath(lc.config.LokiPrefix, u.Path, endpoint)
+
+	if err != nil {
+		return ""
+	}
+
+	if endpoint == "loki/api/v1/tail" {
+		if u.Scheme == "http" {
+			u.Scheme = "ws"
+		} else {
+			u.Scheme = "wss"
+		}
+	}
+
+	return u.String()
+}
+
+func (lc *LokiClient) Ready(ctx context.Context) error {
+	tick := time.NewTicker(500 * time.Millisecond)
+	url := lc.getURLFor("ready", nil)
+	for {
+		select {
+		case <-ctx.Done():
+			tick.Stop()
+			return ctx.Err()
+		case <-lc.t.Dying():
+			tick.Stop()
+			return lc.t.Err()
+		case <-tick.C:
+			lc.Logger.Debug("Checking if Loki is ready")
+			resp, err := http.Get(url)
+			if err != nil {
+				lc.Logger.Warnf("Error checking if Loki is ready: %s", err)
+				continue
+			}
+			_ = resp.Body.Close()
+			if resp.StatusCode != http.StatusOK {
+				lc.Logger.Debugf("Loki is not ready, status code: %d", resp.StatusCode)
+				continue
+			}
+			lc.Logger.Info("Loki is ready")
+			return nil
+		}
+	}
+}
+
+func (lc *LokiClient) Tail(ctx context.Context) (chan *LokiResponse, error) {
+	responseChan := make(chan *LokiResponse)
+	dialer := &websocket.Dialer{} //TODO: TLS support
+	u := lc.getURLFor("loki/api/v1/tail", map[string]string{
+		"limit": strconv.Itoa(lc.config.Limit),
+		"start": strconv.Itoa(int(time.Now().Add(-lc.config.Since).UnixNano())),
+		"query": lc.config.Query,
```

Comment:
it seems the `delay_for` parameter allows to do some magic to deal with log duplication and so on. I suggest we let the user decide how to configure it ?

cf. https://github.com/grafana/loki/issues/931 and https://github.com/grafana/grafana/issues/20346

Reply:
Totally we can add `delay_for` params

Reply:
Development: done
Tests: done